server/controllers/job.controller.js

```diff
@@ -0,0 +1,239 @@
+
+      if (job.isDeleted || (job.expirationDate && job.expirationDate < today)) {
+        return next(new APIError('Not allowed to update this job as it has been deleted', httpStatus.FORBIDDEN));
+      }
+
+      const updated = Object.assign(job, req.body);
+      await updated.save();
+
+      return res.status(httpStatus.OK).json(transform(updated, true));
+    } catch (err) {
+      return next(err);
+    }
+  },
+
+  /**
+  * @swagger
+  *  /jobs/{jobId}:
+  *   get:
+  *     summary: Get job by ID
+  *     description: Get job by ID
+  *     tags: [job]
+  *     security: []
+  *     parameters:
+  *       - $ref: '#/parameters/jobId'
+  *     responses:
+  *       '200':
+  *         description: successful operation
+  *         schema:
+  *           $ref: '#/definitions/Job'
+  *       '404':
+  *         $ref: '#/responses/NotFound'
+  */
+  get: async (req, res, next) => {
+    try {
+      const job = await Job
+        .findById(req.params.jobId);
+
+      if (!job) {
+        return next(new APIError('Job not found', httpStatus.NOT_FOUND));
+      }
+
+      // If the job posting has been logically deleted or expired then it should only
+      // be visible to the user who created it initially
+      const today = new Date().getDate();
+      const isJobAuthor = req.user && (job.postedUser.toString() === req.user._id.toString());
+
+      if (job.isDeleted || (job.expirationDate && job.expirationDate < today)) {
+        if (!req.user) {
+          return next(new APIError('Job not found', httpStatus.NOT_FOUND));
+        }
+
+        if (!isJobAuthor) {
+          return next(new APIError('Job not found', httpStatus.NOT_FOUND));
+        }
+      }
+
+      return res.json(transform(job, isJobAuthor));
+    } catch (err) {
+      return next(err);
+    }
+  },
+
+  /**
+  * @swagger
+  *  /jobs/{jobId}/apply:
+  *   get:
+  *     summary: Applies for a job
+  *     description: Submit a job application to the company who posted the job.
+  *       Should be accompanied by a covering note and a resume.
+  *     tags: [job]
+  *     security:
+  *       Token: []
+  *     parameters:
+  *       - $ref: '#/parameters/jobId'
+  *     requestBody:
+  *       description: A JSON object containing the application details
+  *       content:
+  *         coveringLetter: A short note from the applicant to the employer.
+  *         resume: File containing the applicant's resume to be forwarded to the employer.
+  *     responses:
+  *       '200':
+  *         description: successful operation
+  *         schema:
+  *           $ref: '#/definitions/Job'
+  *       '400':
+  *         description: validation error when required parameters are missing
+  *         $ref: '#/response/BadRequest'
+  *       '404':
+  *         description: The job being applied for does not exist, has been deleted or has expired
+  *         $ref: '#/responses/NotFound'
+  */
+  apply: async (req, res, next) => {
+    try {
```

Comment:
Eventually, it would be nice to log failures for applications somewhere even if it's just to our event-stream system. Since it's the most complex part given it deals with email + attachments.